File: sdk/elixir/test/hooks_test.exs
defmodule RueHook.HooksTest do
  # Dispatch: every request gets a reply carrying its id, every refusal a reason.
  use ExUnit.Case, async: true

  alias RueHook.Hooks

  defp req(kind, op, extra \\ %{}), do: Map.merge(%{"id" => 42, "kind" => kind, "op" => op}, extra)

  test "an op the protocol does not have is refused by name" do
    r = Hooks.answer(%Hooks{}, req("execute", "teleport"))
    assert {r["id"], r["ok"]} == {42, false}
    assert r["error"] =~ "execute.teleport"
  end

  test "an op of an unserved kind is refused, not silent" do
    r = Hooks.answer(%Hooks{}, req("journal", "append", %{"entry" => %{}}))
    assert {r["id"], r["ok"]} == {42, false}
    assert r["error"] =~ "journal.append"
  end

  defmodule Notifier do
    def deliver("warn", _s, _b), do: {:refuse, "paging is off tonight"}
    def deliver("odd", _s, _b), do: :something_else
    def deliver(_l, _s, _b), do: raise("socket closed")
  end

  test "a refusal, a fault and a malformed answer all become reasons" do
    h = %Hooks{notify: Notifier}
    assert Hooks.answer(h, req("notify", "deliver", %{"level" => "warn"}))["error"] == "paging is off tonight"
    fault = Hooks.answer(h, req("notify", "deliver", %{"level" => "err"}))
    assert fault["ok"] == false
    assert fault["error"] =~ "socket closed"
  end

  defmodule Prober do
    def observe(_host, "bare"), do: RueHook.observation(:yes, "a bare value, not {:ok, v}")
    def observe(_host, _), do: {:ok, RueHook.observation(:yes, "ok")}
  end

  test "a handler that returns neither {:ok, v} nor {:refuse, why} is refused, not a crash" do
    # The value fell through to a case with no clause for it, and the
    # CaseClauseError ended the serve loop -- or the embedding host's
    # connection.
    r = Hooks.answer(%Hooks{probe: Prober}, req("probe", "observe", %{"probe" => "bare"}))
    assert {r["id"], r["ok"]} == {42, false}
    assert r["error"] =~ "probe.observe"
  end

  defmodule Facts do
    def read_fact(_host, "file:/present"), do: {:ok, ""}
    def read_fact(_host, _), do: {:ok, nil}
  end

  test "read_fact's no-such-file is an answer with no content" do
    absent = Hooks.answer(%Hooks{execute: Facts}, req("execute", "read_fact", %{"shape" => "file:/absent"}))
    assert absent["ok"] == true
    refute Map.has_key?(absent, "content")
    present = Hooks.answer(%Hooks{execute: Facts}, req("execute", "read_fact", %{"shape" => "file:/present"}))
    assert present["content"] == ""
  end

  test "the registration names only the kinds supplied" do
    reg = RueHook.Serve.registration("x", %Hooks{probe: Prober, notify: Notifier})
    assert reg["kinds"] == ["probe", "notify"]
    assert reg["protocol"] == 1
  end
end

File: sdk/java/src/test/java/dev/rue/hook/HooksTest.java
package dev.rue.hook;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Dispatch: every request gets a reply carrying its id, and every refusal says why. */
class HooksTest {

    private static Map<String, Object> req(String kind, String op, Object... kv) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", 42L);
        m.put("kind", kind);
        m.put("op", op);
        for (int i = 0; i < kv.length; i += 2) {
            m.put((String) kv[i], kv[i + 1]);
        }
        return m;
    }

    @Test
    void anOpTheProtocolDoesNotHaveIsRefusedByName() {
        Map<String, Object> r = new Hooks().answer(req("execute", "teleport"));
        assertEquals(42L, r.get("id"));
        assertEquals(false, r.get("ok"));
        assertTrue(String.valueOf(r.get("error")).contains("execute.teleport"), String.valueOf(r));
        assertEquals(false, new Hooks().answer(req("weather", "report")).get("ok"));
    }

    @Test
    void anOpOfAKindThisHookDoesNotServeIsRefusedNotSilent() {
        Map<String, Object> r = new Hooks().answer(req("journal", "append", "entry", Map.of()));
        assertEquals(42L, r.get("id"));
        assertEquals(false, r.get("ok"));
        assertTrue(String.valueOf(r.get("error")).contains("does not serve journal.append"), String.valueOf(r));
    }

    @Test
    void aHandlersRefusalAndAHandlersFaultBothBecomeReasons() {
        Hooks h = new Hooks();
        h.notify = (level, subject, body) -> {
            if (level.equals("warn")) throw new Hooks.Refusal("paging is off tonight");
            throw new IllegalStateException("socket closed");
        };
        Map<String, Object> refused = h.answer(req("notify", "deliver", "level", "warn"));
        assertEquals("paging is off tonight", refused.get("error"));
        Map<String, Object> fault = h.answer(req("notify", "deliver", "level", "err"));
        assertEquals(false, fault.get("ok"));
        assertTrue(String.valueOf(fault.get("error")).contains("IllegalStateException: socket closed"));
    }

    @Test
    void aReplyMissingARequiredFieldIsRefusedHereNotSentForR0303() {
        Hooks h = new Hooks();
        h.approval = new Hooks.Approval() {
            public List<Object> authenticators() {
                return List.of();
            }

            public String challenge(String i, String d, Object s, Object c) {
                return "c";
            }

            public Map<String, Object> verify(String i, String d, Object s, String a, String p) {
                return Map.of("reason", "forgot the verdict");
            }
        };
        Map<String, Object> r = h.answer(req("approval", "verify"));
        assertEquals(false, r.get("ok"));
        assertTrue(String.valueOf(r.get("error")).contains("without verified"), String.valueOf(r));
    }

    @Test
    void readFactsNoSuchFileIsAnAnswerWithNoContent() {
        Hooks h = new Hooks();
        h.execute = new Hooks.Execute() {
            public Map<String, Object> run(String host, String instance, List<Object> body) {
                return Hooks.output("", Map.of());
            }

            public String readFact(String host, String shape) {
                return shape.endsWith("present") ? "" : null;
            }
        };
        Map<String, Object> absent = h.answer(req("execute", "read_fact", "shape", "file:/absent"));
        assertEquals(true, absent.get("ok"));
        assertFalse(absent.containsKey("content"));
        Map<String, Object> empty = h.answer(req("execute", "read_fact", "shape", "file:/present"));
        assertEquals("", empty.get("content"), "an empty file is not an absent one");
    }

    @Test
    void theRegistrationNamesOnlyTheKindsSupplied() {
        Hooks h = new Hooks();
        h.probe = (host, probe) -> Hooks.observation("yes", "");
        h.notify = (level, subject, body) -> {};
        Map<String, Object> reg = Serve.registration("x", h);
        assertEquals(List.of("probe", "notify"), reg.get("kinds"));
        assertEquals(Op.HOOK_PROTOCOL, reg.get("protocol"));
    }
}
